Pointer expressions

MorMahd opened the pull request:
- Added dedicated pointer expressions.
- Pointers are still encoded like integers.

Commit:
Added rudimentary pointer expressions

Commit:
fixed expr processing

Commit:
ptr correction

Commit:
fixed ptr to int wrong length

Commit:
more fixes related to arch and ptr

Commit:
more fixes and added PtrLiteral

ThomasHaas commented on dartagnan/src/main/java/com/dat3m/dartagnan/expression/pointer/PointerLiteral.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.math.BigInteger;
+
+public final class PointerLiteral extends LiteralExpressionBase<PointerType> {
+
+    private final BigInteger value;
```
I am not sure if we want to have this class. Such a pointer literal can only really refer to a pointer with invalid provenance and would be equivalently represented by an `int2ptr` cast that does not attempt to recover provenance.
Such provenance-less casts can already be expressed by `nullptr + value`, i.e., `ptrlit(value) == int2ptrNoProv(value) == nullptr + value`, with the additional benefit that "value" does not have to be a constant.
Indeed, you could have a helper method `ExpressionFactory.castIntToPtrNoProvenance` that implements this idea and use it for all those cases where we temporarily need to embed an integer into a pointer due to the type system forcing us to (e.g., for implementing `pthread_t` or function pointers in `NaiveDevirtualisation`).
More generally, I could even imagine that we have a `int2PtrWithProv(Expr address, Expr prov)` expression that casts `address` to a pointer with provenance equivalent to that of `prov`, which would yield `int2PtrNoProv(value) == int2PtrWithProv(value, nullptr)`. But for now, this generelization is not needed.

Commit:
visitor changes

Commit:
removed PtrLiteral

Commit:
removed intLiteral

Commit:
Merge remote-tracking branch 'origin/pointerExpressions' into pointerExpressions

Commit:
fixed litmus c

Commit:
more fixes

Commit:
Pointer size and visitor fixes

Commit:
Vulkan modification, tests pass

Commit:
Vulkan modification, all Vulkan tests pass

Commit:
opencl modification

Commit:
Merge remote-tracking branch 'Dat3M/development' into pointerExpressions

Commit:
Merge remote-tracking branch 'Dat3M/development' into pointerExpressions

# Conflicts:
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NaiveDevirtualisation.java

Commit:
dev merge

Commit:
Litmus lkmm changes

Commit:
Litmus c11 fix

Commit:
Added pointer concat and extract

Commit:
Added pointer recognition to ops VisitorOpsConstant

Commit:
--

Commit:
added encoding and printing of pointer tearing

Commit:
fixed test change

Commit:
fixed more pointer compatibility problems

Commit:
added pointer simplification

Commit:
pthread tests pass in 2.timeout

Commit:
fixed some asm problems and ptrCmp refactoring

Commit:
improved pointer casts

Commit:
added pointer addition simplification

Commit:
restructuring of ptr <-> int encoding

Commit:
- improved type recognition in c11 litmus
- deleted ptr size cast
- function-pointer ++

Commit:
fixed litmus type mismatch

Commit:
refactoring

Commit:
Merge branch 'development' into pointerExpressions

# Conflicts:
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorLKMM.java

Commit:
non-termination tests pass

Commit:
fixed the litmus wrong ptr default reg type

Commit:
fixed some litmus compilation problems

Commit:
renaming of some scoped pointer type variables (confusion between pointer and scoped pointer while searching)

Commit:
added comments

Commit:
fixed inline asm unsupported operation

Commit:
fixed litmus c bitvector problem

Commit:
improvement to casts

Commit:
added ptr add to InclusionBasedPointerAnalysis

Commit:
fixes

Commit:
added encoding of ptrAdd in FieldSensitiveAndersen

Commit:
Temp fix for tearing not working correctly. Rll mixed tests pass (still take a long time because alias analysis is not yet optimal)

Commit:
more clarity

Commit:
fixed null ptr store

Commit:
fixed null ptr store

Commit:
small changes

Commit:
merged detached branch

Commit:
mem to reg gep fix to accept ptr add

Commit:
fixed int encoding llvm tests failing

Commit:
fixed one of spirv assertions tests

ThomasHaas committed:
Initial commit for MemoryType support

ThomasHaas committed:
Basic encoding of assignments over memory (WIP)

ThomasHaas committed:
- Added assignEqual methods to model assignments with possible conversions.
- Reworked encodings to use the new assignEqual if the intention is to encode assignments.
- Tearing now uses the new MemoryType to split/concat objects

ThomasHaas committed:
Fix memory cast to types whose size is not a multiple of 8.

ThomasHaas committed:
Minor refactor

ThomasHaas committed:
Add back integer encoding for MemoryExtract

ThomasHaas committed:
Minor refactor
Bug fix in ExprSimplifier for fromMem(toMem(x)) casts

ThomasHaas committed:
Fix buggy initialization of cyclic static dependencies.

ThomasHaas committed:
Cleanup

Commit:
All non broken Litmus tests pass.

Commit:
Fixed memory + 0 and constant ptr mocking

Commit:
Corrected asm riscv ptr type and extended llvm misc test by 4 sec

Commit:
Merge branch 'development' into pointerExpressions

# Conflicts:
#	dartagnan/src/main/java/com/dat3m/dartagnan/expression/type/TypeFactory.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/SparseConditionalConstantPropagation.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ThreadCreation.java

Commit:
.

Commit:
tearing only int inits fix

Commit:
fixed nullptr forced 64 bit encoding

Commit:
fixed llvm pthread test

Commit:
improved pointer casting

Commit:
fixed llvm timeout and mixed tests
refactoring

Commit:
forced litmus riscv ptrs' handling as ints

Commit:
asm bug workaround

Commit:
asm execution status workaround

Commit:
fixed armv8 (armv7 still broken)

Commit:
fixed armv7 asm and null pointer handling problem

Commit:
fixed armv7 asm

Commit:
Merge remote-tracking branch 'origin/pointerExpressions' into pointerExpressions

# Conflicts:
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/memory/MemoryObject.java

Commit:
fixed non constrained variable

Commit:
asm arm pointer casts only when needed

Commit:
fixed some of "others" tests

Commit:
others fixes.

Commit:
fixed more others tests

Commit:
fixed spirv others

Commit:
removed necessary code

Commit:
fixed full alias analysis recursion bug

Commit:
andersen changes

Commit:
Andersen alias fix

Commit:
General improvements

Commit:
Merge branch 'development' into pointerExpressions

# Conflicts:
#	dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/AssumeSolver.java
#	dartagnan/src/test/java/com/dat3m/dartagnan/llvm/MixedTest.java
#	dartagnan/src/test/resources/log4j2.xml

Commit:
General improvements

Commit:
General improvements

Commit:
General improvements

Commit:
Merge remote-tracking branch 'Dat3M/memoryType' into pointerExpressions

# Conflicts:
#	dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExpressionFactory.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExpressionPrinter.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/ExecutionStatus.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/memory/MemoryObject.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Tearing.java

ThomasHaas committed:
Added MemoryExtend expression
Avoid size casts on target type when converting over memory

ThomasHaas committed:
Add actual conversions between memX and bvN for N < X

Commit:
added fixed pointer tearing

Commit:
Merge branch 'memoryType' into pointerExpressions

# Conflicts:
#	dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExpressionFactory.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/expression/type/TypeFactory.java

Commit:
fixed trackability

MorMahd committed:
Merge branch 'development' into pointerExpressions

Commit:
removed redundant code

Commit:
removed forced int ops from riscv

Commit:
removed forced int ops from tso and c11

Commit:
removed forced int binary op

Commit:
removed redundant functions and casts

Commit:
Reformat of modified files

Commit:
Fixed some litmus c tests

Commit:
Removed forced cast in intcmp

Commit:
fix

Commit:
fix

Commit:
clarity and removed unused imports

Commit:
added comments

Commit:
Merge remote-tracking branch 'Dat3M/development' into pointerExpressions

# Conflicts:
#	dartagnan/src/main/java/com/dat3m/dartagnan/GlobalSettings.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/encoding/EncodingContext.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/encoding/PropertyEncoder.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExpressionFactory.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/expression/floats/FloatBinaryExpr.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLlvm.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/program/memory/MemoryObject.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java
#	dartagnan/src/main/java/com/dat3m/dartagnan/wmm/utils/Cut.java

Commit:
added import

Commit:
type fixes

MorMahd commented on dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Inlining.java:
```diff
@@ -142,7 +154,14 @@ private static Event inlineBody(FunctionCall call, Snapshot callTarget, int scop
         for (int j = 0; j < callTarget.parameters.size(); j++) {
             final Register register = registerMap.get(callTarget.parameters.get(j));
-            parameterAssignments.add(newLocal(register, arguments.get(j)));
+            if (register.getType() instanceof PointerType) {
```
should not be resolved here
